File: src/main/java/eu/ec2u/card/handlers/V1.java
/*
 * Copyright © 2020-2022 EC2U Alliance
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package eu.ec2u.card.handlers;

import com.google.inject.Inject;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import eu.ec2u.card.Profile;
import eu.ec2u.card.Profile.Info;
import eu.ec2u.card.Setup;
import eu.ec2u.card.services.*;

import java.io.*;
import java.text.ParseException;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static java.nio.charset.StandardCharsets.UTF_8;

public final class V1 implements HttpHandler {

    private static final Pattern BearerPattern=Pattern.compile("\\s*Bearer\\s*(?<token>\\S*)\\s*");


    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Inject private Setup setup;
    @Inject private Codec codec;
    @Inject private Notary notary;
    @Inject private Fetcher fetcher;


    @Override public void handle(final HttpExchange exchange) throws IOException {

        if ( exchange.getRequestMethod().equals("GET") ) {

            final Profile profile=new Profile()

                    .setInfo(new Info()
                            .setVersion(setup.getVersion())
                            .setInstant(setup.getInstant())
                    );

            final int status=Optional.ofNullable(exchange.getRequestHeaders().getFirst("Authorization"))

                    .map(BearerPattern::matcher)
                    .filter(Matcher::find)
                    .map(matcher -> matcher.group("token"))

                    .flatMap(token -> notary.verify(token))

                    .map(token -> {

                        try {

                            return codec.decode(new StringReader(token), Profile.User.class);

                        } catch ( final IOException e ) {

                            throw new UncheckedIOException(e);

                        } catch ( final ParseException e ) {

                            return null;
                        }

                    })

                    .map(user -> {

                        profile
                                .setUser(user)
                                .setCards(fetcher.fetch(user));  // !!! error handling

                        return 200;

                    })

                    .orElseGet(() -> {

                        exchange.getResponseHeaders().set(
                                "WWW-Authenticate", "Bearer realm=\"card.ec2u.eu"+"\""
                        );

                        return 401;

                    });


            exchange.getResponseHeaders().set("Content-Type", "application/json");

            exchange.sendResponseHeaders(status, 0);

            try (
                    final OutputStream output=exchange.getResponseBody();
                    final Writer writer=new OutputStreamWriter(output, UTF_8)
            ) {

                codec.encode(writer, profile);

            }


        } else {

            exchange.sendResponseHeaders(405, -1);

        }

    }

}

File: code/php/v1.php
<?php
/*
 * Copyright © 2020-2022 EC2U Alliance
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once('esc.php');

$conf = "/etc/ec2u-card/"; // !!! from setup module
$tenants = json_decode(file_get_contents($conf . "tenants.json"), JSON_OBJECT_AS_ARRAY); // !!! exception handling


if (!isset($_SERVER['REMOTE_USER']) || $_SERVER['REMOTE_USER'] == "") {

    header("HTTP/1.1 401 Unauthorized");

} else {

    $user = user();

    if (is_string($user)) {

        header("HTTP/1.1 409 OK");
        header('Content-Type: application/json');

        echo json_encode($user);

    } else {

        $tenant = $tenants[$user["schac"]];

        if (!isset($tenant)) {

            header("HTTP/1.1 409 OK");
            header('Content-Type: application/json');

            echo json_encode("tenant-undefined");

        } else {

            $cards = cards($user["esi"], $tenant);

            if (!is_int($cards)) {

                header("HTTP/1.1 200 OK");
                header('Content-Type: application/json');

                echo json_encode(array(
                    'user' => $user,
                    'cards' => $cards
                ));

            } else if ($cards == 500) {

                header("HTTP/1.1 500 Internal Server Error");

            } else if ($cards == 502) {

                header("HTTP/1.1 502 Bad Gateway");

            }

        }
    }

}


function user()
{

    $user = array();

    if (!isset($_SERVER["cn"]) || $_SERVER["cn"] == "") {
        return "edugain-undefined-cn";
    }

    $user["name"] = $_SERVER["cn"];

    if (!isset($_SERVER["mail"]) || $_SERVER["mail"] == "") {
        return "edugain-undefined-mail";
    }

    $user["email"] = $_SERVER["mail"];

    if (!isset($_SERVER["schacPersonalUniqueCode"]) || $_SERVER["schacPersonalUniqueCode"] == "") {
        return "edugain-undefined-schacPersonalUniqueCode";
    }

    $user["esi"] = $_SERVER["schacPersonalUniqueCode"];

    if (!isset($_SERVER["schacHomeOrganisation"]) || $_SERVER["schacHomeOrganisation"] == "") {
        return "edugain-undefined-schacHomeOrganisation";
    }

    $user["schac"] = $_SERVER["schacHomeOrganisation"];

    // !!! eduPersonScopedAffiliation

    return $user;

}


?>
